pkg/ingester/retention.go

```diff
@@ -39,183 +66,363 @@ func defaultRetentionPolicy() retentionPolicy {
+func (dc *diskCleaner) DeleteUploadedBlocks(ctx context.Context) int {
+	tenantIDs, err := dc.blockManager.GetTenantIDs(ctx)
 	if err != nil {
-		return nil, err
+		level.Error(dc.logger).Log(
+			"msg", "failed to delete uploaded blocks, could not read tenant ids",
+			"err", err,
+		)
+		return 0
 	}
-	var blockDirs []fs.DirEntry
-	for _, tenantDir := range tenants {
-		if !tenantDir.IsDir() {
+
+	var deleted int
+	for _, tenantID := range tenantIDs {
+		blocks, err := dc.blockManager.GetBlocksForTenant(ctx, tenantID)
+		if err != nil {
+			level.Error(dc.logger).Log(
+				"msg", "failed to delete uploaded blocks, could not get blocks for tenant",
+				"err", err,
+				"tenantID", tenantID,
+			)
 			continue
 		}
-		tenantID := tenantDir.Name()
-		tenantDirPath := filepath.Join(dir, tenantID, phlareDBLocalPath)
-		if blockDirs, err = fs.ReadDir(e.fileSystem, tenantDirPath); err != nil {
-			if os.IsNotExist(err) {
-				// Must be created by external means, skipping.
+
+		expiryTs := time.Now().Add(-dc.config.BlockExpiry)
+		for _, block := range blocks {
+			if !block.Uploaded {
 				continue
 			}
-			return nil, err
-		}
-		for _, blockDir := range blockDirs {
-			if !blockDir.IsDir() {
+
+			if ulid.Time(block.ID.Time()).After(expiryTs) {
 				continue
 			}
-			blockPath := filepath.Join(tenantDirPath, blockDir.Name())
-			if blockID, ok := block.IsBlockDir(blockPath); ok {
-				blocks = append(blocks, &tenantBlock{
-					ulid:     blockID,
-					path:     blockPath,
-					tenantID: tenantID,
-				})
+
+			err = dc.blockManager.DeleteBlock(ctx, block)
+			switch {
+			case os.IsNotExist(err):
+				level.Warn(dc.logger).Log(
+					"msg", "failed to delete uploaded block, does not exist",
+					"err", err,
+					"path", block.Path,
+				)
+			case err != nil:
+				level.Error(dc.logger).Log(
+					"msg", "failed to delete uploaded block",
+					"err", err,
+					"path", block.Path,
+				)
+			default:
+				deleted++
 			}
-			// A malformed/invalid ULID likely means that the
-			// directory is not a valid block, ignoring.
 		}
 	}
-
-	// Sort the blocks by their id, which will be the time they've been created.
-	sort.Slice(blocks, func(i, j int) bool {
-		return blocks[i].ulid.Compare(blocks[j].ulid) < 0
-	})
-
-	return blocks, nil
+	return deleted
 }
 
-func (e *retentionPolicyEnforcer) cleanupBlocksWhenHighDiskUtilization(ctx context.Context) error {
-	var volumeStatsPrev *diskutil.VolumeStats
-	volumeStatsCurrent, err := e.volumeChecker.HasHighDiskUtilization(e.dbConfig.DataPath)
+// EnforceHighDiskUtilization will run more aggressive disk cleaning if high
+// disk utilization is detected. It returns true if high disk utilization was
+// detected, along with the number of files deleted and the estimated bytes
+// recovered. If no high disk utilization was detected, false is returned.
+func (dc *diskCleaner) EnforceHighDiskUtilization(ctx context.Context) (int, int, bool) {
```

Comment:
Nitpick, the name suggests to me that the method tries to enforce "high" utilization. Perhaps `diskCleaner.EnforceDiskUtilization` or `diskCleaner.ReduceDiskUtilization` could make it more clear?

Reply:
I agree. Fixed in https://github.com/grafana/pyroscope/pull/2816/commits/82a1a229196f0c4cb094bf384c930d25d59b6f54